### PR TITLE
Fix DH_check() excessive time with over sized modulus (1.1.1)

### PR DESCRIPTION
**Pull request by mattcaswell**
The DH_check() function checks numerous aspects of the key or parameters
that have been supplied. Some of those checks use the supplied modulus
value even if it is excessively large.

There is already a maximum DH modulus size (10,000 bits) over which
OpenSSL will not generate or derive keys. DH_check() will however still
perform various tests for validity on such a large modulus. We introduce a
new maxium (32,768) over which DH_check() will just fail.

An application that calls DH_check() and supplies a key or parameters
obtained from an untrusted source could be vulernable to a Denial of
Service attack.

The function DH_check() is itself called by a number of other OpenSSL
functions. An application calling any of those other functions may
similarly be affected. The other functions affected by this are
DH_check_ex() and EVP_PKEY_param_check().

CVE-2023-3446 (low severity)

### CODE REVIEW

**Comment by mattcaswell** [comment]
This is a backport of #21451 for the 1.1.1 branch

File: include/openssl/dh.h [hunk 1]
```diff
@@ -29,6 +29,9 @@ extern "C" {
 # ifndef OPENSSL_DH_MAX_MODULUS_BITS
 #  define OPENSSL_DH_MAX_MODULUS_BITS    10000
 # endif
+# ifndef OPENSSL_DH_MAX_CHECK_MODULUS_BITS
+#  define OPENSSL_DH_CHECK_MAX_MODULUS_BITS  32768
```
Comment (tom-cosgrove-arm) [inline include/openssl/dh.h:33]: Should it be `OPENSSL_DH_MAX_CHECK_MODULUS_BITS` or `OPENSSL_DH_CHECK_MAX_MODULUS_BITS`?

Comment (mattcaswell) [reply include/openssl/dh.h:33]: Fixed

**Comment by mattcaswell** [comment]
Updated to address comments (both from here and the master version). Please take another look.

**Comment by FdaSilvaYY** [comment]
Typo in commit message: maxium

**Review by paulidale — APPROVED** [review]
With the typo fix.

**Comment by mattcaswell** [comment]
Rebased to fix commit message typo.

File: CHANGES [hunk 2]
```diff
@@ -9,7 +9,21 @@
...
+    The function DH_check() performs various checks on DH parameters. One of those
+    checks confirms that the modulus ("p" parameter) is not too large. Trying to use
+    a very large modulus is slow and OpenSSL will not normally use a modulus which
+    is over 10,000 bits in length.
+
+    However the DH_check() function checks numerous aspects of the key or parameters
+    that have been supplied. Some of those checks use the supplied modulus value
+    even if it has already been found to be too large.
+
+    A new limit has been added to DH_check of 32,768 bits. Supply a key/parameters
```
Comment (tom-cosgrove-arm) [inline CHANGES:23]: Again "Supply" on its own sounds a bit clunky, but not major

**Review by tom-cosgrove-arm — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by tom-cosgrove-arm** [comment]
Approval still stands - thanks

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Review by bernd-edlinger — APPROVED** [review]

**Comment by t8m** [comment]
Merged to 1.1.1 branch. Thank you.

**Pull request closed without merging**